lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -1671,6 +1671,13 @@ LogicalResult FIRRTLLowering::run() {
     }
   }
 
+  // Replace all backedges with uses of their regular values.
+  for (auto &[backedge, value] : backedges) {
+    while (backedges.count(value))
+      value = backedges[value];
```

Comment:
I guess this is the source of hang. To prevent infinite loop, how about limiting the number of iterations to be less than `backedges.size()`? (it is a bit hacky way to detect cycles though)

Reply:
Huh yeah good point, this is pretty nasty. My suggestion would be to add a `SmallSet<Value> seen;` just above and add something like `&& seen.insert(value)` to the while condition. That way we would just stop as soon as we close the loop.

But on the other hand, something tells me that this should never create an infinite loop, even if the backedges describe a combinational loop themselves. We should still be able to resolve the backedges even if there's a combinational loop. We might be missing a `backedges[value] = resolvedValue;` when we resolve a backedge :thinking:. Or we should just remove the entry from the `backedges` map as soon as the backedge is resolved.

Reply:
I think (hope) I can fix this by making it look more like classic equiv classes, will see what I can do.

Reply:
This now handles the undriven combinational loops by creating an undriven wire and using that as the value, which is the same as what we do for ports with no driver.  All the code I've written for this looks almost identical to the `createTmpWire` stuff in this pass, albeit a bit more powerful since everything connected together but undriven will share the same wire.  I think we can follow up with this change by replacing all uses of `createTmpWire` with `createBackedge`.

Reply:
This is awesome! Replacing `createTmpWire` with `createBackedge` sounds like it would get rid of temporary wires in almost all cases, which sounds fantastic! 🎉